File: algorithm-design/leetcode/lc017_letter_combinations_of_a_phone_number/lc017_letter_combinations_of_a_phone_number.py
#!/usr/bin/env python3
# https://leetcode.com/problems/letter-combinations-of-a-phone-number/

import unittest

from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if digits is None or len(digits) == 0:
            return []
        mapping = [
            ' ',
            '1', 'abc', 'def',
            'ghi', 'jkl', 'mno',
            'pqrs', 'tuv', 'wxyz']
        mapped = []
        mapped.append('')
        for digit in digits:
            index = ord(digit) - ord('0')
            longer = []
            for prefix in mapped:
                for suffix in mapping[index]:
                    longer.append(prefix + suffix)
            mapped = longer
        return mapped


class TestCode(unittest.TestCase):
    def test_empty(self):
        result = Solution().letterCombinations('')
        self.assertEqual(len(result), 0)

    def test_example(self):
        expected = ['ad', 'ae', 'af', 'bd', 'be', 'bf', 'cd', 'ce', 'cf']
        expected = sorted(expected)
        result = Solution().letterCombinations('23')
        result = sorted(result)
        self.assertEqual(result, expected)

File: algorithm-design/leetcode/lc017_letter_combinations_of_a_phone_number/LC017LetterCombinationsOfAPhoneNumber.cs
// https://leetcode.com/problems/letter-combinations-of-a-phone-number/

using System;
using System.Collections.Generic;

namespace AlgorithmDesign.leetcode.lc017_letter_combinations_of_a_phone_number
{
    public class Solution
    {
        public IList<string> LetterCombinations(string digits)
        {
            if (String.IsNullOrEmpty(digits)) return new List<string>();
            string[] mapping = { " ", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
            IList<string> mapped = new List<string>();
            mapped.Add("");
            foreach (char digit in digits)
            {
                int index = digit - '0';
                List<string> longer = new List<string>();
                foreach (string prefix in mapped)
                {
                    foreach (char suffix in mapping[index])
                    {
                        longer.Add(prefix + suffix);
                    }
                }
                mapped = longer;
            }
            return mapped;
        }
    }
}
